### PR TITLE
Create README.md

### PR DESCRIPTION
**Pull request by Bartmax**
First iteration. Modifications and suggestions are welcome.
fix for https://github.com/openiddict/core/issues/1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# OpenIddict, as an OpenID Connect Addict
+
+### What's Open Iddict ?
```
Comment (kevinchalet) [inline README.md]: nit: OpenIddict.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+OpenIddict library aims to provide a simple and easy out of the box solution to implement your OpenID Connect 
+server for ASP.NET 5.
+OpenIddict leverages the use of Identity (for user management) and Entityframework (as store provider).
+
+### What's Open ID Connect ?
```
Comment (kevinchalet) [inline README.md]: nit: OpenID Connect.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+Open ID is a protocol that allows multiple clients to delegate the Authentication to one or multiple Authority servers.
```
Comment (kevinchalet) [inline README.md]: nit: `Open ID` -> `OpenID Connect` (OpenID and OpenID Connect are two different specs)

Comment (kevinchalet) [inline README.md]: nit: `the Authentication` -> `authentication` (or `the authentication process`)

File: README.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+Facebook, Google, Twitter, are all well-known as OpenId Authority Servers.
+
+### Why should I want/need my own Open ID Connect Server?
```
Comment (kevinchalet) [inline README.md]: nit: OpenID Connect server

File: README.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+There are many scenarios on when this may be useful, 
+
+ 1. Having multiple applications with one central place for authentication.
+ 2. Having non-trusted clients (like JavaScript, Desktop or Mobile applications).
```
Comment (kevinchalet) [inline README.md]: Actually, the client are not necessarily non-trusted. You can have totally trusted MVC clients.
I guess you meant external clients, where you want to allow third-party developers to access your users' data?

File: README.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+ 3. Needs or wants to use single sign-on for multiple services.
```
Comment (kevinchalet) [inline README.md]: Sounds like the same as the first one, no?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+In general terms, you want to have a centralized authentication server/login and then any allowed clients
```
Comment (kevinchalet) [inline README.md]: This sentence sounds a bit... weird? :smile:

File: README.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+## Getting Started
+
+To use OpenIddict Server you need to include OpenIddict as a dependency in your project.json:
+```
+    "dependencies": {
+        "OpenIddict": "1.0.0-*"
+    },
+```
+
+In ConfigureServices there's a handy extension method for IdentityBuilder that you need to set up.
```
Comment (kevinchalet) [inline README.md]: nit: consider using backticks for types/methods/properties: ConfigureServices -> `ConfigureServices`

File: README.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        // options
+    });
+}
+```
+
+> `UseOpenIddict()` must be **AFTER** `app.UseIdentity()` and any external providers.
+
+### Configuration & Options.
+
+<p>OpenIddict have multiple options you can set to customize for your requirements/needs.<p>
```
Comment (kevinchalet) [inline README.md]: Has?

File: README.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    The URI must be absolute and may contain a path, but no query string or fragment part.<br>
+    Unless AllowInsecureHttp has been set to true, an HTTPS address must be provided.
+</td>
+<td><code>oidc-server</code></td>
+</tr>
+<tr>
+  <td>Issuer</td>
+  <td>Uri</td>
+  <td>The base address (absolute) used to uniquely identify the authorization server.<br>
+      <sub><i>Unless <code>AllowInsecureHttp</code> has been set to <code>true</code>, an HTTPS address must be provided.</i></sub>
```
Comment (kevinchalet) [inline README.md]: I should probably fix ASOS' documentation to mention this property is now optional. When left to null, the issuer property is automatically inferred from the request URL.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+## Support
+
+**Need help or wanna share your thoughts? Don't hesitate to join our dedicated chat rooms:**
+
+- **JabbR: [https://jabbr.net/#/rooms/aspnet-contrib](https://jabbr.net/#/rooms/aspnet-contrib)**
+
+## Contributors
+
+**OpenIddict** is actively maintained by **[Kévin Chalet](https://github.com/PinpointTownes)**. Contributions are welcome and can be submitted using pull requests.
```
Comment (kevinchalet) [inline README.md:102]: Feel free to add your name! :clap:

File: README.md (hunk 2 above)
Comment (ilmax) [inline README.md]: nit: link to OpenID connect site/documentation

File: README.md [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        .AddDbContext<ApplicationDbContext>(options =>
+            options.UseSqlServer(Configuration["Data:DefaultConnection:ConnectionString"]));
+
+    services.AddIdentity<ApplicationUser, IdentityRole>()
+        .AddEntityFrameworkStores<ApplicationDbContext>()
+        .AddDefaultTokenProviders()
+        .AddOpenIddict();
+
+    /// .... other services you may have
+}
```
Comment (ilmax) [inline README.md]: missing closing ```

File: README.md (hunk 2 above)
Comment (Bartmax) [inline README.md]: Great point @ilmax ! Should we mention ASOS too and link to it ?

Comment (ilmax) [inline README.md]: Sure :)

**Comment by kevinchalet** [comment]
Should we also indicate that nightly builds can now be found on the aspnet-contrib MyGet repository? https://www.myget.org/F/aspnet-contrib/api/v3/index.json

File: README.md [hunk 13]
```diff
@@ -0,0 +1,224 @@
+# OpenIddict, as an OpenID Connect Addict
+
+### What's OpenIddict ?
+
+OpenIddict library aims to provide a simple and easy out of the box solution 
+to implement an OpenID Connect server for ASP.NET 5.
+
+It does leverages the use of Identity (for user management) and
```
Comment (kevinchalet) [inline README.md]: nit: `does leverages` -> `leverages`.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+Having multiple applications; maintainance of user databases and 
+registration/login workflows can easily became an administrative and 
+security nightmare.
+
+The established solution to these problems is to delegate user authentication 
+and provisioning to a dedicated, purpose-built service, called an Identity 
+Provider (IdP).
+
+OpenID Connect is a standard protocol on top of OAuth that enables Clients to 
+verify the identity of the End-User based on the authentication performed by
```
Comment (ilmax) [inline README.md]: on top of OAuth 2.0 protocol

File: README.md [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+    <li>Signin.cshtml</li>
+    <li>Error.cshtml</li>
+  </ul>
+  </td>
+  <td><code>false</code></td>
+</tr>
+</tbody>
+</table>
+
+You can find working samples in the [samples](https://github.com/openiddict/core/tree/dev/samples) directory.
```
Comment (kevinchalet) [inline README.md]: I think we should move these 2 paragraphs in the `Getting started` section.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+Nightly builds can now be found on the [aspnet-contrib](https://github.com/aspnet-contrib) MyGet repository https://www.myget.org/F/aspnet-contrib/api/v3/index.json.
```
Comment (kevinchalet) [inline README.md]: Yeah, move this one at the very begging of the `Getting started` section :smile:

File: README.md [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+  <td>bool</td>
+  <td>
+    Set to <code>true</code> to allow incoming requests to arrive on HTTP and to allow `redirect_uri` parameters to have HTTP URI addresses.
+    <br>
+    <sub>Setting this option to <code>false</code> in production is strongly encouraged to mitigate man-in-the-middle attacks.</sub>
+  </td>
+  <td><code>false</code></td>
+</tr>
+<tr>
+  <td>UseCustomViews</td>
```
Comment (kevinchalet) [inline README.md]: FYI: this option will be removed when https://github.com/openiddict/core/pull/25 is merged :warning:

Comment (Bartmax) [inline README.md]: :+1: :watch:

File: README.md [hunk 18]
```diff
@@ -0,0 +1,212 @@
+# OpenIddict, as an OpenID Connect Addict
```
Comment (kevinchalet) [inline README.md]: OpenIddict, the OpenID Connect server you'll be addict to :smile:

File: README.md [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+
+### What's OpenIddict ?
+
+OpenIddict library aims to provide a simple and easy out of the box solution
```
Comment (kevinchalet) [inline README.md]: nit: `OpenIddict library aims to provide` -> `OpenIddict aims at providing`

File: README.md [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+
+### What's OpenIddict ?
+
+OpenIddict library aims to provide a simple and easy out of the box solution 
+to implement an OpenID Connect server for ASP.NET 5.
+
+It leverages the use of Identity (for user management) and 
+EntityFramework (as a store provider).
+
+Under the hood it uses [AspNet.Security.OpenIdConnect.Server](https://github.com/PinpointTownes/AspNet.Security.OpenIdConnect.Server)
```
Comment (kevinchalet) [inline README.md]: The sentence sounds weird (and the link should point to aspnet-contrib, not to my fork :smile:)

File: README.md [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+### What's OpenID Connect ?
+
+Applications often need to identify their users. 
+The simplistic approach is to create a local database for the users’ accounts 
+and credentials. While this may work well for some scenarios, people find
+signup and account creation to be tedious which translates to less users using
+your application.
+
+Having multiple applications; maintainance of user databases and 
+registration/login workflows can easily became an administrative and
```
Comment (kevinchalet) [inline README.md]: nit: `became` -> `become`

File: README.md [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+
+For more documentation, visit [OpenID Connect web site](http://openid.net/connect/)
+
+### Why an OpenID Connect Server?
+
+A consumer web site can greatly streamline user onboarding by integrating 
+login with existing Identity Providers.
+On the enterprise side, this would be ideally be one internal Identity Provider.
+
+Having an internal Idenity Provider provides you ability to use local password
```
Comment (kevinchalet) [inline README.md]: typo: `Idenity`

File: README.md [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+
+### Why an OpenID Connect Server?
+
+A consumer web site can greatly streamline user onboarding by integrating 
+login with existing Identity Providers.
+On the enterprise side, this would be ideally be one internal Identity Provider.
+
+Having an internal Idenity Provider provides you ability to use local password 
+authentication, control the information that is exposed to the client applications
+and more generally, you can control who access your API.
```
Comment (kevinchalet) [inline README.md]: `access` -> `can access`

File: README.md [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+        // options
+    });
+}
+```
+
+> **Note:** `UseOpenIddict()` must be used ***after*** `app.UseIdentity()` and any external providers.
+
+### Configuration & Options.
+
+OpenIddict have multiple options you can set to customize for your
```
Comment (kevinchalet) [inline README.md]: `have` -> `has`

**Comment by PureKrome** [comment]
Thoughts (in order of reading the read.me) ... and of course, this is IMO ...

### Title
- Title header is too long. Split into two.
- Loose the emoji in the title (or sub-heading if you do that).

### Under the title - first quick summary
- Needs `dev` and `master` badges. Ie. nuget version and nuget download count (for each). MyGet for `dev` branch. (then use AppVeyor for CI/CD).
- Needs a screenie for the NuGet cli: `install-package hello.is.it.me.youre.looking.for-.6.9.691

### Sales pitch
- Too long.
- I would literally try and keep the first sentence as the main paragraph describing the service. 
- Split the other para's into a others .. like _Why would I use this_ and _What happens under the hood_"

### What's OpenID Connect ?
- I think this is not really important (just yet) and should be dropped lower in the readme.
- Reword: "_The simplistic approach is to create a local database for the users’ accounts and credentials._" to (something like) "_Most developers usually create a `Users` database table and have a `UserName` or `Email`, `Password` and (hopefully) `PasswordSalt` fields for storing a clients credentials._"
- This whole section feels ... a bit long. But maybe it's necessary though.

### Why an OpenID Connect Server?
- Ah! i jumped the gun a bit (above) with suggesting this to be earlier. I still stand by that, thought.
- This entire section made no sense to me :( I still don't _get it_. Some hard examples are needed here. For example: Website sells movie tickets. Accounts are needed to purchase a ticket, blah blah blah. OpenIdDict can offer this and do that.. blah blah. money shot!

### All the rest below this.
- The rest (IMO) are wiki items. Too much for a ReadMe. Move all the code stuff into a wiki with pages .. not a page, but pages...

### Support / Contrib / Lic
- perfect. end the read me these :)

### Questions I'm wondering about, but not answered?
- OK, so this is some website thingy. Can i use this in my own _existing_ website? or is this a stand alone thing?
- If i can use it in my existing website, what about my existing DB? will it `fuck.things.up`:tm: ?

**Comment by kevinchalet** [comment]
@PureKrome thanks for this great feedback! :clap: 

@Bartmax if you want to add the AppVeyor/Travis badges in this PR:

AppVeyor:

```
[![Build status](https://ci.appveyor.com/api/projects/status/d0d8git3o6lqkvbm?svg=true)](https://ci.appveyor.com/project/aspnet-contrib/core)
```

Travis:

```
[![Build status](https://travis-ci.org/openiddict/core.svg)](https://travis-ci.org/openiddict/core)
```

**Comment by kevinchalet** [comment]
@Bartmax do you want to include the recommendations made by @PureKrome before we merge this PR? :stuck_out_tongue:

**Comment by Bartmax** [comment]
ill make at least some in the next 2 days. let's wait for them before merge.

File: README.md [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+
+### What's OpenIddict ?
+
+OpenIddict aims at providing a simple and easy out of the box solution 
+to implement an OpenID Connect server for ASP.NET 5.
+
+
+### Why an OpenID Connect Server?
+
+With an OpenID Conenct Server you can manage all your users using local
```
Comment (kevinchalet) [inline README.md]: There's a little typo here: `OpenID Conenct Server` -> `OpenID Connect Server`.

File: README.md [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+the information that is exposed to each client. 
+
+
+### How does it work?
+
+OpenIddict, by default, leverages the use of Identity (for user management) and 
+EntityFramework (as an optional store provider).
+
+Adding it to your existing application allows you to register clients and 
+serve them authenticate tokens.
```
Comment (kevinchalet) [inline README.md:28]: `authenticate tokens` -> `authentication tokens`?

File: README.md [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+
+--------------
+
+## Getting Started
+
+You can find working samples in the [samples](https://github.com/openiddict/core/tree/dev/samples) directory.
+
+Nightly builds can now be found on the [aspnet-contrib](https://github.com/aspnet-contrib) MyGet repository https://www.myget.org/F/aspnet-contrib/api/v3/index.json.
+
+To use OpenIddict Server you need to include OpenIddict as a dependency in your project.json:
```
Comment (kevinchalet) [inline README.md:43]: `OpenIddict Server`?

File: README.md [hunk 28]
```diff
@@ -0,0 +1,106 @@
+OpenIddict<br><sub>The OpenID Connect server you'll be addict to.</sub>
+=======
+
+AppVeyor: [![Build status](https://ci.appveyor.com/api/projects/status/d0d8git3o6lqkvbm?svg=true)](https://ci.appveyor.com/project/aspnet-contrib/core)
+
+Travis: [![Build status](https://travis-ci.org/openiddict/core.svg)](https://travis-ci.org/openiddict/core)
+
```
Comment (PureKrome) [inline README.md:7]: Include current NuGet version badges.

Comment (kevinchalet) [inline README.md:7]: FYI, there's no official release on NuGet.org yet. I'd really like to upload an alpha version for RC1 though.

File: README.md [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+### What's OpenIddict ?
+
+OpenIddict aims at providing a simple and easy out of the box solution 
+to implement an OpenID Connect server for ASP.NET 5.
+
+
+### Why an OpenID Connect Server?
+
+With an OpenID Connect Server you can manage all your users using local 
+password or an external identity provider management for all your applications
```
Comment (PureKrome) [inline README.md:17]: Current: `or an external identity provider management`
Suggestion: `or an external identity provider (e.g. Login with Facebook) for all your...`

File: README.md [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+to implement an OpenID Connect server for ASP.NET 5.
+
+
+### Why an OpenID Connect Server?
+
+With an OpenID Connect Server you can manage all your users using local 
+password or an external identity provider management for all your applications 
+in one central place, with the power to control who can access your API and 
+the information that is exposed to each client. 
+
```
Comment (PureKrome) [inline README.md:20]: A screen shot here might be worth 'a money shot'...

File: README.md [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+
+### How does it work?
+
+OpenIddict, by default, leverages the use of Identity (for user management) and
```
Comment (PureKrome) [inline README.md:24]: Link `Identity` and `EF` to their respective GH repo's.

File: README.md (hunk 26 above)
Comment (PureKrome) [inline README.md:28]: `authenticate` tokens or `authenticaTION` tokens? (please don't throw stones at me if I've got this wrong, i always mess up my words...)

File: README.md [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+
+OpenIddict, by default, leverages the use of Identity (for user management) and 
+EntityFramework (as an optional store provider).
+
+Adding it to your existing application allows you to register clients and 
+serve them authenticate tokens.
+
+Under the hood it uses [AspNet.Security.OpenIdConnect.Server](https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server) 
+middleware that works with any standards-compliant OAuth 2.0/OpenID Connect 
+client including the official OpenID Connect client middleware
```
Comment (PureKrome) [inline README.md:32]: Add link to the GH repo...

File: README.md (hunk 27 above)
Comment (PureKrome) [inline README.md:43]: `To use....`  whoa... hold up a sec.

I think this needs to have a list summary of the steps needed to use this nuget package, before any code and text.

eg.

Steps to setup your webapp
1. install-package
2. modify `json.settings`
3. Wire up middleware in `ConfigureServices` (in `startup.cs`)

and now we can go into some code, etc...

Bascially, i need to get an overall feel about how complex it is (via the number of steps), to use this in my app.

Looking at the readme text below this point, i'm not sure if this is _everything_ or just the _hardest_ parts, of the steps.

Remember, i'm blond..... i need to be spoon fed.

File: README.md (hunk 30 above)
Comment (kevinchalet) [inline README.md:20]: The native UI needs a large rework (it's the "basic" UI I used for ASOS samples):

image

**Comment by kevinchalet** [comment]
Merged, with the changes suggested by @PureKrome: https://github.com/openiddict/core/commit/ae38ba119360ae2d30f34f15bb952d55793a9a49

@PureKrome thanks for the review! :clap:

**Pull request closed without merging**